File: src/main/java/xyz/devvydont/smprpg/gui/items/MenuTrashItems.java
package xyz.devvydont.smprpg.gui.items;

import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.format.TextDecoration;
import org.bukkit.Sound;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.event.inventory.InventoryOpenEvent;
import org.jetbrains.annotations.NotNull;
import xyz.devvydont.smprpg.SMPRPG;
import xyz.devvydont.smprpg.gui.base.MenuBase;
import xyz.devvydont.smprpg.services.ItemService;
import xyz.devvydont.smprpg.util.formatting.ComponentUtils;
import xyz.devvydont.smprpg.util.formatting.Symbols;

public final class MenuTrashItems extends MenuBase {
    public MenuTrashItems(@NotNull Player player) {
        super(player, 6);
    }

    @Override
    protected void handleInventoryOpened(InventoryOpenEvent event) {
        // Prepare the inventory
        event.titleOverride(ComponentUtils.merge(
                ComponentUtils.create(Symbols.OFFSET_NEG_32 + Symbols.OFFSET_NEG_16 + Symbols.TRASH_MENU, NamedTextColor.WHITE),
                ComponentUtils.create(Symbols.OFFSET_NEG_128 + Symbols.OFFSET_NEG_32 + Symbols.OFFSET_NEG_8 + Symbols.OFFSET_NEG_4 + "Trash Items", NamedTextColor.BLACK)));
        this.setMaxStackSize(100);

        // Render the UI
        this.clear();
    }

    @Override
    protected void handleInventoryClicked(InventoryClickEvent event) {
        event.setCancelled(false);

        if (event.getCurrentItem() == null)
            return;

        var itemBlueprint = SMPRPG.getService(ItemService.class).getBlueprint(event.getCurrentItem());
        // If the item clicked is enchanted or reforged, we should prevent the item from being shift clicked.
        if (event.isShiftClick() && (!event.getCurrentItem().getEnchantments().isEmpty() || itemBlueprint.isReforged(event.getCurrentItem()))) {
            var error = ComponentUtils.merge(ComponentUtils.create("WHOA THERE!", NamedTextColor.RED, TextDecoration.BOLD), ComponentUtils.create(" Do you really want to trash this? Manually drag the item in the menu if so."));
            player.sendMessage(ComponentUtils.alert(error, NamedTextColor.RED));
            player.playSound(player.getLocation(), Sound.ENTITY_VILLAGER_DEATH, 1f, 1.5f);
            event.setCancelled(true);
        }
    }

    @Override
    protected void handleInventoryClosed(InventoryCloseEvent event) {
        var itemCount = 0;
        for (var itemStack : this.getItems()) {
            if (itemStack == null) {
                continue;
            }

            itemCount += itemStack.getAmount();
            itemStack.setAmount(0);
        }

        this.player.sendMessage(ComponentUtils.success(ComponentUtils.merge(
            ComponentUtils.create("You trashed ", NamedTextColor.GREEN),
            ComponentUtils.create(String.valueOf(itemCount), NamedTextColor.AQUA),
            ComponentUtils.create(" items!", NamedTextColor.GREEN)
        )));
    }
}

File: src/main/java/xyz/devvydont/smprpg/gui/items/MenuTrashItems.kt
package xyz.devvydont.smprpg.gui.items

import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.Sound
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.inventory.InventoryOpenEvent
import xyz.devvydont.smprpg.SMPRPG
import xyz.devvydont.smprpg.gui.base.MenuBase
import xyz.devvydont.smprpg.services.ItemService
import xyz.devvydont.smprpg.util.formatting.ComponentUtils
import xyz.devvydont.smprpg.util.formatting.Symbols

class MenuTrashItems(player: Player) : MenuBase(player, 6) {
    override fun handleInventoryOpened(event: InventoryOpenEvent) {
        // Prepare the inventory
        event.titleOverride(
            ComponentUtils.merge(
                ComponentUtils.create(
                    Symbols.OFFSET_NEG_32 + Symbols.OFFSET_NEG_16 + Symbols.TRASH_MENU,
                    NamedTextColor.WHITE
                ),
                ComponentUtils.create(
                    Symbols.OFFSET_NEG_128 + Symbols.OFFSET_NEG_32 + Symbols.OFFSET_NEG_8 + Symbols.OFFSET_NEG_4 + "Trash Items",
                    NamedTextColor.BLACK
                )
            )
        )
        this.setMaxStackSize(100)

        // Render the UI
        this.clear()
    }

    override fun handleInventoryClicked(event: InventoryClickEvent) {
        event.setCancelled(false)

        if (event.getCurrentItem() == null) return

        val itemBlueprint =
            SMPRPG.getService<ItemService?>(ItemService::class.java)!!.getBlueprint(event.getCurrentItem()!!)
        // If the item clicked is enchanted or reforged, we should prevent the item from being shift clicked.
        if (event.isShiftClick() && (!event.getCurrentItem()!!.getEnchantments().isEmpty() || itemBlueprint.isReforged(
                event.getCurrentItem()
            ))
        ) {
            val error = ComponentUtils.merge(
                ComponentUtils.create("WHOA THERE!", NamedTextColor.RED, TextDecoration.BOLD),
                ComponentUtils.create(" Do you really want to trash this? Manually drag the item in the menu if so.")
            )
            player.sendMessage(ComponentUtils.alert(error, NamedTextColor.RED))
            player.playSound(player.getLocation(), Sound.ENTITY_VILLAGER_DEATH, 1f, 1.5f)
            event.setCancelled(true)
        }
    }

    override fun handleInventoryClosed(event: InventoryCloseEvent) {
        var itemCount = 0
        for (itemStack in this.items) {
            if (itemStack == null) {
                continue
            }

            itemCount += itemStack.getAmount()
            itemStack.setAmount(0)
        }

        this.player.sendMessage(
            ComponentUtils.success(
                ComponentUtils.merge(
                    ComponentUtils.create("You trashed ", NamedTextColor.GREEN),
                    ComponentUtils.create(itemCount.toString(), NamedTextColor.AQUA),
                    ComponentUtils.create(" items!", NamedTextColor.GREEN)
                )
            )
        )
    }
}
